source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+`nsInfo` array of the namespace on which the operation should be performed as its value. The
+documents have the following format:
+
+#### Insert
+
+```json
+{
+    "insert": <Int32>,
+    "document": <Document>
+}
+```
+
+Drivers MUST add an `_id` field at the beginning of the insert document if one is not already
+present.
+
+When a user executes a bulk write with an unacknowledged write concern, drivers SHOULD check the
+size of the insert document to verify that it does not exceed `maxBsonObjectSize`. For acknowledged
+bulk writes, drivers MAY rely on the server to return an error if the document exceeds
+`maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's
+`hello` response.
+
+#### Update
+
+```json
+{
+    "update": <Int32>,
+    "filter": <Document>,
+    "updateMods": <Document | Array>,
+    "multi": <Boolean>,
+    "upsert": Optional<Boolean>,
+    "arrayFilters": Optional<Array>,
+    "hint": Optional<String>
+}
+```
+
+#### Delete
+
+```json
+{
+    "delete": <Int32>,
+    "filter": <Document>,
+    "multi": <boolean>,
+    "hint": Optional<String>,
+    "collation": Optional<Document>
+}
+```
+
+### Namespace Information
+
+The `nsInfo` field is an array containing the namespaces on which the write operations should be
+performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the
+`nsInfo` array have the following format:
+
+```json
+{
+    "ns": <String>
+}
+```
+
+### `errorsOnly` and `verboseResults`
+
+The `errorsOnly` field indicates whether the results cursor returned in the `bulkWrite` response
+should contain only errors and omit individual results. If false, both individual results for
+successful operations and errors will be returned.
+
+This field corresponds to the `verboseResults` option defined on `BulkWriteOptions`. Its value
+should be provided as `false` if `verboseResults` was set to `true`; otherwise, it should be
+provided as `true`. This field MUST always be defined in the `bulkWrite` command regardless of
+whether the user specified a value for `verboseResults`.
+
+## Command Batching
+
+Drivers MUST accept an arbitrary number of operations as input to the `MongoClient.bulkWrite` method.
+Because the server imposes restrictions on the size of write operations, this means that a single
+call to `MongoClient.bulkWrite` may require multiple `bulkWrite` commands to be sent to the server.
+Drivers MUST split bulk writes into separate commands when the user's list of operations exceeds
+one or more of these maximums: `maxWriteBatchSize`, `maxBsonObjectSize`, and `maxMessageSizeBytes`.
+Each of these values can be retrieved from the selected server's `hello` command response. Drivers
+MUST merge results from multiple batches into a single `BulkWriteResult` or `BulkWriteException`
+to return from `MongoClient.bulkWrite`.
+
+### Number of Writes
+
+The `ops` array MUST NOT include more than `maxWriteBatchSize` operations.
```

Comment:
Given the third prose test, I assume clients do not enforce this and instead rely on the server to do so. Should that be clarified here? I don't recall seeing it mentioned earlier.